open_spiel/games/2048.cc

```diff
@@ -0,0 +1,383 @@
+    return absl::StrCat(std::to_string(chance_action.is_four ? 4 : 2), 
+        " added to row ", std::to_string(chance_action.row + 1),
+        ", column ", std::to_string(chance_action.column + 1));
+  }
+  switch (action_id) {
+    case kMoveUp:
+      return "Up";
+      break;
+    case kMoveRight:
+      return "Right";
+      break;
+    case kMoveDown:
+      return "Down";
+      break;
+    case kMoveLeft:
+      return "Left";
+      break;
+    default:
+      return "Invalid action";
+      break;
+  }  
+}
+
+int TwoZeroFourEightState::AvailableCellCount() const {
+  int count = 0;
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      if (BoardAt(r, c).value == 0) {
+        count++;
+      }
+    }
+  }
+  return count;
+}
+
+ActionsAndProbs TwoZeroFourEightState::ChanceOutcomes() const {
+  ActionsAndProbs action_and_probs;
+  int count = AvailableCellCount();
+  if (count == 0) {
+    action_and_probs.reserve(1);
+    action_and_probs.emplace_back(kNoCellAvailableAction, 1);
+    return action_and_probs;  
+  }
+  action_and_probs.reserve(count * 2);
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      if (BoardAt(r, c).value == 0) {
+        // 2 appearing randomly on the board should be 9 times as likely as a 4
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, false)), .9 / count);
+        action_and_probs.emplace_back(ChanceActionToSpielAction(
+            ChanceAction(r, c, true)), .1 / count);
+      }      
+    }
+  }  
+  return action_and_probs;
+}
+
+std::vector<Action> TwoZeroFourEightState::LegalActions() const {
+  if (IsTerminal()) {
+    return {};
+  }
+  if (IsChanceNode()) {
+    return LegalChanceOutcomes();
+  }
+  return kPlayerActions;
+}
+
+bool TwoZeroFourEightState::InBounds(int row, int column) const {
+  return (row >= 0 && row < kDefaultRows && column >= 0
+      && column < kDefaultColumns);
+}
+
+std::string TwoZeroFourEightState::ToString() const {  
+  std::string str;
+  for (int r = 0; r < kDefaultRows; ++r) {
+    for (int c = 0; c < kDefaultColumns; ++c) {
+      std::string tile = std::to_string(BoardAt(r, c).value);
+      absl::StrAppend(&str, std::string(5 - tile.length(), ' '));
+      absl::StrAppend(&str, tile);
+    }
+    absl::StrAppend(&str, "\n");
+  }
+  return str;  
+}
+
+bool TwoZeroFourEightState::IsTerminal() const {
+  return Reached2048() 
+      || (AvailableCellCount() == 0 && !TileMatchesAvailable());
+}
+
+bool TwoZeroFourEightState::Reached2048() const {
+  for (int r = 0; r < kDefaultRows; r++) {
+    for (int c = 0; c < kDefaultColumns; c++) {
+      if (BoardAt(r, c).value == 2048) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+std::vector<double> TwoZeroFourEightState::Returns() const {
+  if (IsTerminal()) {
+    if (Reached2048()) {
+      return {1.0};
+    }
```

Comment:
Ah yes, I see now, you're treating it as a win/loss game. In this case, we should probably be consistent with how the RL community has treated this environment.

Reply:
Have changed it from a win/loss game to match the scoring system of the original game